Framework/MDAlgorithms/src/Integrate3DEvents.cpp

```diff
@@ -69,6 +72,266 @@ void Integrate3DEvents::addEvents(
   }
 }
 
+std::pair<boost::shared_ptr<const Geometry::PeakShape>,
+          std::tuple<double, double, double>>
+Integrate3DEvents::integrateStrongPeak(const IntegrationParameters &params,
+                                       const V3D &peak_q, double &inti,
+                                       double &sigi) {
+
+  inti = 0.0; // default values, in case something
+  sigi = 0.0; // is wrong with the peak.
+  auto result = getEvents(peak_q);
+  if (!result)
+    return std::make_pair(boost::make_shared<NoShape>(),
+                          make_tuple(0., 0., 0.));
+
+  const auto &events = result.get();
+  if (events.empty())
+    return std::make_pair(boost::make_shared<NoShape>(),
+                          make_tuple(0., 0., 0.));
+
+  DblMatrix cov_matrix(3, 3);
+  makeCovarianceMatrix(events, cov_matrix, params.regionRadius);
+
+  std::vector<V3D> eigen_vectors;
+  getEigenVectors(cov_matrix, eigen_vectors);
+
+  std::vector<double> sigmas;
+  for (int i = 0; i < 3; i++) {
+    sigmas.push_back(stdDev(events, eigen_vectors[i], params.regionRadius));
+  }
+
+  bool invalid_peak =
+      std::any_of(sigmas.cbegin(), sigmas.cend(), [](const double sigma) {
+        return std::isnan(sigma) || sigma <= 0;
+      });
+
+  if (invalid_peak)
+    return std::make_pair(boost::make_shared<NoShape>(),
+                          make_tuple(0., 0., 0.));
+
+  const auto max_sigma = *std::max_element(sigmas.begin(), sigmas.end());
+  if (max_sigma == 0)
+    return std::make_pair(boost::make_shared<NoShape>(),
+                          make_tuple(0., 0., 0.));
+
+  auto rValues = calculateRadiusFactors(params, max_sigma);
+  auto &r1 = std::get<0>(rValues), r2 = std::get<1>(rValues),
+       r3 = std::get<2>(rValues);
+
+  std::vector<double> abcBackgroundOuterRadii, abcBackgroundInnerRadii;
+  std::vector<double> peakRadii;
+  for (int i = 0; i < 3; i++) {
+    abcBackgroundOuterRadii.push_back(r3 * sigmas[i]);
+    abcBackgroundInnerRadii.push_back(r2 * sigmas[i]);
+    peakRadii.push_back(r1 * sigmas[i]);
```

Comment:
Could be. This is a more general issue with the Integrate3DEvents class which in general expects 3D stuff but uses std::vector (because it was probably written pre-c++11)